### PR TITLE
quantized transpose operator

### PR DESCRIPTION
**Pull request by RafLit**
## Description ##
added quantized transpose operator

### CODE REVIEW

**Review by bartekkuncer — APPROVED** [review]
Looks good.

File: src/operator/quantization/dnnl/dnnl_quantized_transpose.cc [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  return DNNLStorageType(attrs, dev_mask, true, dispatch_mode, in_attrs, out_attrs);
+}
+
+static void DNNLQuantizedTransposeForward(const nnvm::NodeAttrs& attrs,
+                                          const OpContext& ctx,
+                                          const std::vector<NDArray>& inputs,
+                                          const std::vector<OpReqType>& req,
+                                          const std::vector<NDArray>& outputs) {
+  CHECK(inputs[0].dtype() == mshadow::kUint8 || inputs[0].dtype() == mshadow::kInt8)
+      << "dnnl_quantized_pooling op only supports uint8 and int8 as input type";
```
Comment (agrabows) [inline src/operator/quantization/dnnl/dnnl_quantized_transpose.cc:47]: This comment (dnnl_quantized_pooling) does not fit in here.

File: src/operator/quantization/quantized_transpose.cc [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    .add_argument("data", "NDArray-or-Symbol", "A ndarray/symbol of type `float32`")
+    .add_argument("min_data",
+                  "NDArray-or-Symbol",
+                  "The minimum scalar value "
+                  "possibly produced for the data")
+    .add_argument("max_data",
+                  "NDArray-or-Symbol",
+                  "The maximum scalar value "
+                  "possibly produced for the data")
+    .add_arguments(TransposeParam::__FIELDS__());
```
Comment (agrabows) [inline src/operator/quantization/quantized_transpose.cc:95]: Did you consider adding FQuantizable attribute with _return QuantizeType::kSupport;_ ?

File: tests/python/quantization/test_quantization.py [hunk 3]
```diff
@@ -713,6 +713,32 @@ def forward(self, x):
         check_quantized_fc((256, 2048, 2, 2), 800, False, qdtype)
         check_quantized_fc((256, 111, 2, 2), 800, False, qdtype)
 
+@use_np
+def test_quantized_transpose():
```
Comment (agrabows) [inline tests/python/quantization/test_quantization.py:717]: Maybye add some test for quantized transpose after some other quantizable operator e.g. conv, fc.

File: src/operator/quantization/quantized_transpose.cc [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+NNVM_REGISTER_OP(_contrib_quantized_transpose)
+    .add_alias("_npx_quantized_transpose")
+    .set_num_inputs(3)
+    .set_num_outputs(3)
+    .set_attr_parser(ParamParser<TransposeParam>)
+    .set_attr<mxnet::FInferShape>("FInferShape", QuantizedTransposeShape)
+    .set_attr<nnvm::FInferType>("FInferType", QuantizedTransposeType)
+    // TODO(Xinyu): a temp solution to enable GluonCV INT8 flow,
```
Comment (agrabows) [inline src/operator/quantization/quantized_transpose.cc:68]: Is this comment valid for this change?

File: src/operator/quantization/dnnl/dnnl_quantized_transpose.cc [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  if (req[0] == kNullOp) {
+    return;
+  }
+  CHECK_EQ(inputs.size(), 3U);
+  CHECK_EQ(outputs.size(), 3U);
+  DNNLRun(DNNLTransposeForward<TransposeParam>, attrs, ctx, inputs[0], req[0], outputs[0]);
```
Comment (agrabows) [inline src/operator/quantization/dnnl/dnnl_quantized_transpose.cc:53]: Shouldn't we check if inputs pass through _SupportDNNLTranspose_ functions' conditions before calling _DNNLTransposeForward_?

File: src/operator/quantization/quantized_transpose.cc [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        })
+    .set_attr<nnvm::FListOutputNames>(
+        "FListOutputNames",
+        [](const NodeAttrs& attrs) {
+          return std::vector<std::string>{"output", "min_output", "max_output"};
+        })
+    .set_attr<nnvm::FInplaceOption>(
+        "FInplaceOption",
+        [](const NodeAttrs& attrs) {
+          return std::vector<std::pair<int, int> >{{0, 0}, {1, 1}, {2, 2}};
```
Comment (agrabows) [inline src/operator/quantization/quantized_transpose.cc:84]: Do we use inplace of inputs[0] and outputs[0] for Transpose operator?

File: src/operator/quantization/quantized_transpose.cc [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        })
+    .add_argument("data", "NDArray-or-Symbol", "A ndarray/symbol of type `float32`")
```
Comment (agrabows) [inline src/operator/quantization/quantized_transpose.cc:86]: Isn't data for this operator only int8 and uint8?

File: src/operator/quantization/dnnl/dnnl_quantized_transpose.cc (hunk 5 above)
Comment (bartekkuncer) [reply src/operator/quantization/dnnl/dnnl_quantized_transpose.cc:53]: I believe Rafał wrote the checks here as they are different than the ones in SupportDNNLTranspose function. I agree that they could be moved to a function.

File: src/operator/quantization/quantized_transpose.cc (hunk 4 above)
Comment (bartekkuncer) [reply src/operator/quantization/quantized_transpose.cc:68]: I believe this comment is taken from another operator and as its scope is bigger than just transpose op, checking whether it is still valid could be done in a separate task.

File: src/operator/quantization/dnnl/dnnl_quantized_transpose.cc (hunk 5 above)
Comment (agrabows) [reply src/operator/quantization/dnnl/dnnl_quantized_transpose.cc:53]: Fair point. To be precise it concerns me that in similar function _TransposeComputeExCPU()_ we call _SupportDNNLTranspose()_ function with e.g. _data.shape().Size() == 0_ condition before calling _DNNLRun_. If this and other checks are necessary I believe we should include them here as well. If not probably we should exclude them from _SupportDNNLTranspose()_ in other PR.

File: src/operator/quantization/quantized_transpose.cc (hunk 4 above)
Comment (agrabows) [reply src/operator/quantization/quantized_transpose.cc:68]: If so I think this issue can be resolved.

File: src/operator/quantization/dnnl/dnnl_quantized_transpose.cc [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+      << "dnnl_quantized_transpose only supports uint8 and int8 as input type";
+  if (req[0] == kNullOp) {
+    return;
+  }
+  CHECK_EQ(inputs.size(), 3U);
+  CHECK_EQ(outputs.size(), 3U);
+  if (SupportDNNLQuantizedTranspose(inputs[0])) {
+    DNNLRun(DNNLTransposeForward<ParamType>, attrs, ctx, inputs[0], req[0], outputs[0]);
+  } else {
+    FallBackCompute(UnaryOp::IdentityCompute<cpu>, attrs, ctx, inputs, req, outputs);
```
Comment (bgawrych) [inline src/operator/quantization/dnnl/dnnl_quantized_transpose.cc:66]: Why it is fallback to IdentityCompute - I think it should be fallback to NumpyTranspose<cpu> / Transpose<cpu>

File: tests/python/dnnl/subgraphs/test_conv_subgraph.py [hunk 9]
```diff
@@ -72,6 +72,26 @@ def forward(self, x):
...
+
+    def forward(self, x):
+      out = self.conv0(x)
+      out = mx.np.transpose(out, axes=[0,1,3,2])
+      out = self.conv1(out)
+      return out
+
+  attr = {'conv': []}
+  net = Conv_Transpose_Conv()
+  check_fusion(net, data_shape, attr)
```
Comment (bgawrych) [inline tests/python/dnnl/subgraphs/test_conv_subgraph.py:94]: I think in this case check_quantize call should be used - probably you can check if transpose was quantized by adding it's name to attr dict

**Review by bgawrych — CHANGES_REQUESTED** [review]

**Comment by RafLit** [comment]
@mxnet-bot run ci [website]

**Comment by RafLit** [comment]
@mxnet-bot run ci [windows-gpu, website]

**Comment by RafLit** [comment]
@mxnet-bot run ci [website]

**Comment by RafLit** [comment]
@mxnet-bot run ci [website]

**Comment by RafLit** [comment]
@mxnet-bot run ci [website]

**Comment by RafLit** [comment]
@mxnet-bot run ci [website]

**Comment by RafLit** [comment]
@mxnet-bot run ci [website]

**Comment by RafLit** [comment]
@mxnet-bot run ci [website]

**Review by agrabows — APPROVED** [review]
LGTM

File: src/operator/quantization/quantized_transpose.cc [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+            return std::vector<std::string>{"data", "min_data", "max_data"};                 \
+          })                                                                                 \
+      .set_attr<nnvm::FListOutputNames>(                                                     \
+          "FListOutputNames",                                                                \
+          [](const NodeAttrs& attrs) {                                                       \
+            return std::vector<std::string>{"output", "min_output", "max_output"};           \
+          })                                                                                 \
+      .set_attr<FQuantizable>("FQuantizable",                                                \
+                              [](const NodeAttrs& attrs) { return QuantizeType::kSupport; }) \
+      .add_argument("data", "NDArray-or-Symbol", "Array to be reshaped.")                    \
```
Comment (bgawrych) [inline src/operator/quantization/quantized_transpose.cc:86]:
```suggestion
      .add_argument("data", "NDArray-or-Symbol", "Array to be transposed.")                    \
```

**Comment by RafLit** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by RafLit** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by RafLit** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by RafLit** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by RafLit** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by RafLit** [comment]
@mxnet-bot run ci [unix-cpu]

File: src/operator/quantization/dnnl/dnnl_quantized_transpose.cc [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+ */
+
+/*!
+ * \file dnnl_quantized_transpose.cc
+ * \author: Rafal Litka, rafal.litka@intel.com
+ */
+#if MXNET_USE_ONEDNN == 1
+#include "../../numpy/np_matrix_op-inl.h"
+#include "../../tensor/matrix_op-inl.h"
+#include "../../nn/dnnl/dnnl_transpose-inl.h"
```
Comment (bgawrych) [inline src/operator/quantization/dnnl/dnnl_quantized_transpose.cc:28]:
```suggestion
#include "operator/numpy/np_matrix_op-inl.h"
#include "operator/tensor/matrix_op-inl.h"
#include "operator/nn/dnnl/dnnl_transpose-inl.h"
```

some time ago PR was merged to replace relative paths in dnnl files

**Comment by RafLit** [comment]
@mxnet-bot run ci [website]

**Review by bgawrych — APPROVED** [review]

**Pull request merged**